test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py

```diff
@@ -0,0 +1,148 @@
+    ld = LaunchDescription(actions)
+    ls = LaunchService(debug=True)
+    ls.include_launch_description(ld)
+
+    loop = osrf_pycommon.process_utils.get_loop()
+    launch_task = loop.create_task(ls.run_async())
+    loop.run_until_complete(asyncio.sleep(timeout_sec))
+    if not launch_task.done():
+        loop.create_task(ls.shutdown())
+        loop.run_until_complete(launch_task)
+    assert 0 == launch_task.result()
+    return ls.context
+
+
+def _create_node_container(*, parameters=None, remappings=None, namespace=''):
+    return ComposableNodeContainer(
+        package='rclcpp_components',
+        executable='component_container',
+        name=TEST_CONTAINER_NAME,
+        namespace=namespace,
+        output='screen',
+        parameters=parameters,
+        remappings=remappings,
+    )
+
+
+def _load_composable_node(*, parameters=None, remappings=None):
+    return LoadComposableNodes(
+        target_container=TEST_CONTAINER_NAME,
+        composable_node_descriptions=[
+            ComposableNode(
+                package='composition',
+                plugin='composition::Listener',
+                name=TEST_NODE_NAME,
+                parameters=parameters,
+                remappings=remappings,
+            )
+        ])
+
+
+def test_load_invalid_node():
+    """Test loading an invalid node."""
+    load_composable_nodes = LoadComposableNodes(
+        target_container=TEST_CONTAINER_NAME,
+        composable_node_descriptions=[
+            ComposableNode(
+                package='nonexistent_package',
+                plugin='this_plugin_should_not_exist',
+                name=TEST_NODE_NAME,
+            )
+        ])
+    context = _assert_launch_no_errors([_create_node_container(), load_composable_nodes])
+
+    assert get_node_name_count(context, f'/{TEST_NODE_NAME}') == 0
+    assert get_node_name_count(context, f'/{TEST_CONTAINER_NAME}') == 1
+
+
+def test_load_node():
+    """Test loading a node."""
+    context = _assert_launch_no_errors([
+        _create_node_container(), _load_composable_node()
+    ])
+
+    assert get_node_name_count(context, f'/{TEST_NODE_NAME}') == 1
+    assert get_node_name_count(context, f'/{TEST_CONTAINER_NAME}') == 1
+
+
+def test_load_node_with_global_remaps_in_group():
+    """Test loading a node with global remaps scoped to a group."""
+    load_composable_nodes_action = _load_composable_node()
+    context = _assert_launch_no_errors([
+        _create_node_container(),
+        GroupAction(
+            [
+                SetRemap('chatter', 'new_topic_name'),
+                load_composable_nodes_action,
+            ],
+            scoped=True,
+        ),
+    ])
+
+    assert get_node_name_count(context, f'/{TEST_NODE_NAME}') == 1
+    assert get_node_name_count(context, f'/{TEST_CONTAINER_NAME}') == 1
+
+    # Check the remaps in load service request
+    assert len(load_composable_nodes_action._LoadComposableNodes__load_node_requests) == 1
```

Comment:
maybe the sent request could be a property, so we don't have to check a private member variable

Reply:
I followed a similar pattern that is used in [test_node.py](https://github.com/ros2/launch_ros/blob/9250313932c2c7477a982ef0c0cfe5124d3cfddf/test_launch_ros/test/test_launch_ros/actions/test_node.py#L91).

I think a user should have access to that information; it seems like an implementation detail subject to change.

Reply:
> I followed a similar pattern that is used in test_node.py.

Yeah, it's not a great pattern, but acceptable I guess.

> I think a user should have access to that information; it seems like an implementation detail subject to change.

That sounds contradictory, I guess you meant "should not".

---

I prefer not testing implementation details, but I don't feel strongly in this case.

Reply:
Yeah, sorry, I meant "should not" (updated my comment).

Reply:
I guess the ideal way to write tests in this situation would be to write a mock component container and assert things about it (e.g. if the correct service request was received). It adds a bit more to the test complexity, but could be better long-term. What do you think?

Reply:
> What do you think?

Sounds like a good proposal to me, though I don't mind approving this as-is.

Reply:
I went ahead and refactored the tests to use a mock object. I think it's much nicer. PTAL 5d9dcf3